### ISSUE TITLE
MatTooltip: should change detection be triggered by showing a tooltip?

### ISSUE THREAD

**Issue by victornoel** [angular/material2#10306]
#### Bug, feature request, or proposal:

Bug

#### What is the expected behavior?

Hovering an element with a `matTooltip` should not trigger change detection on this element if it is an `OnPush` element and its `@Input()`s does not change.

#### What is the current behavior?

Hovering an element with a `matTooltip` triggers change detection on this element.

This only happens if `matTooltip` is defined using `[matTooltip]="..."` and not when it is defined using `matTooltip="..."`

#### What are the steps to reproduce?

See https://stackblitz.com/edit/angular-material2-issue-ocg9sj
 - pass the mouse on the red `div`
 - you get logs everytime the function needed to draw the component are called
 - you don't get any log from  `ngOnChanges`
 - I would expect that since the `@Input()` didn't change, the component shouldn't be redraw!

#### What is the use-case or motivation for changing an existing behavior?

Prevent useless redraws

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?

cf stackblitz

**Comment by c69** [comment]
Run into this today - not only the hover causes the host element to re-run change detection, but in my case whole outer `*ngFor` is updated...

Please consider increasing priority to P2. Because this essentially means you cannot use matTooltip in a data grid.

P.s.: removed `matTooltip` from our code for now, in favor of native `title`.

**Comment by chihuahua** [comment]
This makes `matTooltip` uninstantaneous to use when change detection happens to take a long time. We've decided to use the title attribute for now, but will keep monitoring this bug.

**Comment by pranavjain** [comment]
I am also not able to use matTooltip when ChangeDetectionStrategy is OnPush.
@chihuahua: Were you able to find any fix of the same yet?

**Comment by gamsterx** [comment]
I just ran into the same problem. I guess we have to live with the `title` attribute for now. This doesn't seem to be very high on the priority list.

**Comment by tsheils** [comment]
I have run into this with a component injected into a portal inside a table cell (complicated enough?) When I hover over the paginator buttons, showing the tooltip, the injected component is destroyed and re-injected.

**Comment by Need-To-Learn** [comment]
Hi, I'm having this issue too with matTooltip directive.
The component's changeDetection is set to ChangeDetectionStrategy.OnPush.
On hover / leave, the component's change detection is triggered multiple times.

![mattooltip_cd](https://user-images.githubusercontent.com/23190578/58703514-3a292f00-83a9-11e9-95e3-342fab318acf.gif)

Thx !

**Comment by klinki** [comment]
I agree, this is a problem for bigger applications with slow change detection cycle. It brings us a lot of problems :(

**Comment by yard2010** [comment]
Hi, is this being working on right now? Can I try to fix it and add a PR?

**Comment by pycraft114** [comment]
@andrewseguin any recent news about this?

**Comment by Affuta** [comment]
It has been 2 years since this issue has been opened and still not resolved...
Is there any plan for change detection to removed for the mat tooltip's hover state? 
This is such a nasty side effect and causing such performance issues throughout our entire applications. 
Has anyone been able to come up with some sort of alternative solution? 
Or is that solution to remove mat tooltip module?

**Comment by yakimko** [comment]
I found out if remove the second param from ComponentPortal https://github.com/angular/components/blob/5284a57b2f24319359d0528522feeca9c1393751/src/material/tooltip/tooltip.ts#L336

a problem will disappear.

do we really need to set it?

**Comment by Toso82** [comment]
Hello all i have the same and use matTooltip="..." as @c69 not solve. I add than use all in latest version

**Comment by jmeinlschmidt** [comment]
Issue is still occurring (@angular/material 11.2.2) :-( any updates?

**Comment by ronen009** [comment]
i just wasted 3 hours on this.
also, changed from matTooltip to title.

https://stackblitz.com/edit/angular-ctmbfe?file=src/app/table-basic-example.html

**Comment by ronen009** [comment]
same happens when hovering over 
mat-sort-header

**Comment by PabloLema** [comment]
Any solution? I have the same issue.

**Comment by Frazer** [comment]
Is Yakimko's suggestion a fix?
https://github.com/angular/components/issues/10306#issuecomment-627378769

or perhaps it is the beginning of a fix, to make the _portal persist?

**Comment by carboncs16** [comment]
Any solution for this as I am facing the same performance issue

**Comment by PhilippMeissner** [comment]
The only feasible "solution" we found so far is wrapping the tooltip into a dumb component which in returns implements the tooltip. This limits the affected components re-running due to a change-detection..

**Comment by Frazer** [comment]
We also had to create our own tooltip component, as found no way to make this work.

**Comment by powhound4** [comment]
4 years later and still no fix for this??

**Comment by nhleto** [comment]
Came across this today while dealing with the same issue

**Comment by nhleto** [comment]
Came across this today while dealing with the same issue

**Comment by PhilippMeissner** [comment]
> This is quite a painful bug. With a large list contained within an OnPush CdkVirtualFor (utilising trackBy) to only render the viewable items, on _every_ tooltip show() call, the pure pipes to retrieve the tooltip content for _every_ single item in the list are retriggered.

I guess your only feasible option will be to wrap the content containing the pipe in another component to minimize the affected components that will get triggered for re-rendering.
Or we wait for another 4 years until this issue eventually becomes interesting enough.

**Comment by sujeetjaiswara** [comment]
primeng tooltip is the best alternative 🔥

**Comment by SeregPie** [comment]
https://github.com/angular/components/issues/10306#issuecomment-627378769

Thank you. ❤️ 

I had only one tooltip in my wrapper component. So this is the workaround I came up with.

```ts
@ViewChild(MatTooltip)
set matTooltip(v: MatTooltip) {
  delete (v as any)._viewContainerRef;
}
```

**Comment by box-turtle** [comment]
https://github.com/angular/components/issues/10306#issuecomment-1206204298 is genius

**Comment by abudygold** [comment]
I successfully created a custom tooltip with HTML, used MouseOver to call the method, and changed innerHTML by ClassName. 

    <mat-icon
    				class="tw-cursor-pointer"
    				#tooltip="matTooltip"
    				matTooltip="sample text"
    				[matTooltipPosition]="tooltipPosition"
    				(mouseover)="onHover()">
    				info
    			</mat-icon>

....

    onHover(): void {
    		setTimeout(() => {
    			document.getElementsByClassName('mdc-tooltip__surface')[0].innerHTML =
    				`<span class="tw-font-bold">This is a tooltip</span> <br/><br/>`;
    		});
    	}
